src/jetstream.rs

```diff
@@ -731,33 +664,692 @@ impl JetStream {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "deliver subject can not be set for an ordered consumer",
+                ));
+            }
+        }
+
+        let process_consumer_info = |info: ConsumerInfo| {
+            // Make sure this new subject matches or is a subset.
+            if !info.config.filter_subject.is_empty() && subject != info.config.filter_subject {
+                return Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "subject does not match consumer",
+                ));
+            }
+
+            if let Some(deliver_group) = info.config.deliver_group.as_ref() {
+                if let Some(queue) = maybe_queue {
+                    if deliver_group != queue {
+                        return Err(io::Error::new(
+                            io::ErrorKind::Other,
+                            format!(
+                                "cannot create a queue subscription {} for a consumer with a deliver group {}",
+                                queue, deliver_group
+                            ),
+                        ));
+                    }
+                } else {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!(
+                            "cannot create a subscription for a consumer with a deliver group {}",
+                            deliver_group
+                        ),
+                    ));
+                }
+            } else {
+                // Prevent an user from attempting to create a queue subscription on
+                // a consumer that was not created with a deliver group.
+                if maybe_queue.is_some() {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "cannot create a queue subscription for a consumer without a deliver group",
+                    ));
+                }
+
+                // Need to reject a non queue subscription to a non queue consumer if the consumer
+                // is already bound.
+                if maybe_queue.is_some() && info.push_bound {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        "consumer is already bound to a subscription",
+                    ));
+                }
+            }
+
+            // Check for configuration mismatches.
+            if let Some(options) = maybe_options {
+                if options.durable_name.is_some()
+                    && options.durable_name != info.config.durable_name
+                {
+                    return Err(io::Error::new(
+                        io::ErrorKind::Other,
+                        format!("configuration requests durable name to be {:?}, but consumer's value is {:?}", options.durable_name, info.config.durable_name
```

Comment:
I don't like our usage of `io::Error` in general, but it shows it weaknesses especially with those parametrized error, all of kind `Other`.

This (params) will make it impossible to check for those explicit errors, though on other hand, that additional information is very valuable for the user.

Reply:
Agree, but we should take a pass and strongly type all the errors everywhere for this to be useful.

Reply:
What would be the alternative. Our own errors yes?

Reply:
Which ever way we go, a simple example with main and the ability to use '?' is desired.

Reply:
Yeah it would be to strongly type every single error that is being returned across the entire crate.
Should be considered but punting from this pull request.